### DISCUSSION TITLE
请问milvus filtering的实现机制？

### DISCUSSION THREAD

**Discussion by xiaobingxia-at** [milvus-io/milvus#38645] [Q&A and General discussion]
最近在公司内部部署milvus，内部有提问，从大体量数据 search while filtering的过程中，milvus是如何保证搜索质量和搜索延迟的？
能否简单解释一下这个filtering metadata的机制？是：
1. 从index搜top-n，然后filter，如果不够了，继续搜下一个page的top-n，再filter。直到最后的结果满足了top-n个数据点？
2. 还是其他优化方式？

除此之代，大家也想了解一下，用partition key。通过search filter by partition key and metadata的整个过程。
1. 每个partition key对应的partition有自己的index吗？（尽管很多partition key定义的逻辑partition最终是划在一个物理partition中）
2. partition key定义后的collection，数据分布是怎么样的呢？（比如，一个物理分区下有很多逻辑分区，每个逻辑分区下有自己的segment，每个segment有自己的index？还是其他分布方式）
2. 能不能search by "filtering by multiple partition key values"?
谢谢。

**Comment by xiaobingxia-at** [comment]
如果有发表的文章介绍实现机制就最好了。谢谢。

**Comment by yhmo** [comment]
原理很简单，主要原理是“先过滤，后搜索”，"bitset机制"
https://milvus.io/docs/bitset.md#Overview
https://milvus.io/docs/filtered-search.md#Filtered-Search

假设有5条数据：
{id:1, vector1}
{id:2, vector2}
{id:3, vector3}
{id:4, vector4}
{id:5, vector5}

建立一个具有5个标志位的bitset: [1, 1, 1, 1, 1]
如果你search with filter "id > 3"，那么就把这个bitset设置为 [0, 0, 0, 1, 1]
然后搜索的时候就只计算后边两条，前3条直接skip
如果topk=2，那么就正好两条数据返回。如果topk=3，也只返回两条，因为符合条件的就两条。如果topk=1，那就把两条结果中最近的那条返回。

=======================================================================
partition无非就是把数据从物理上隔离，每个partition的数据放在一个单独的文件夹下
partition key无非就是自动划分partition，搜索时自动找到parititon

每个collection里有若干partiiton，每个partition里有若干个segment，每个segment有单独的index

打个比方，一个collection想当于一个学校，一个partition相当于一个年级，一个segment想当于一个班

==================================================================
"filtering by multiple partition key values" --------------------------- 当然可以，就好比我想从某个学校里的2年级和4年级这两个年级里找到姓刘的小学生。

**Reply by xiaobingxia-at** [reply]
非常感谢您的详细解释。

> 搜索的时候就只计算后边两条，前3条直接skip

如果这是个brute force的过程，很好理解。但假如是从vector index出发该怎么理解呢？一个vector index是针对一整个segment搞的吧，他也是针对这个segment搜索进行train的。如果很多数据点提前被过滤掉了，依然从这个segment的index出发，进行搜索这一小部分数据。会影响搜索结果吗？（recall rate会降低吗？）

**Reply by xiaobingxia-at** [reply]
> "filtering by multiple partition key values" --------------------------- 当然可以

这可能会导致在多个物理partition上进行搜索吧。这内部是怎么处理的呢？
是在每个partition上分别搜索，然后对结果进行merge吗？（如果结果数量不够，还要在各个分区继续返回下一个page的搜索结果？）

**Reply by yhmo** [reply]
有index的条件下，有的数据被skip掉，就有可能造成结果集不足topk条。在开源的IVF索引和HNSW索引都会有这种问题。在zilliz cloud上的Cardinal索引里由于做了特殊处理，就不会有这种问题，Cardinal的结果集不足时会扩大搜索范围。

每个segment各搜出topk条，n个segment就有n * topk条，这些都是按distance排好序的，很容易就能得到最终topk。很简单，每个高中高考前十名，排好序的，很容易得到全市的前十名。

**Reply by xiaobingxia-at** [reply]
> 有index的条件下，有的数据被skip掉，就有可能造成结果集不足topk条。在开源的IVF索引和HNSW索引都会有这种问题。在zilliz cloud上的Cardinal索引里由于做了特殊处理，就不会有这种问题，Cardinal的结果集不足时会扩大搜索范围。

所以在结果集不足topk条时，可以通过iterator继续搜索下一page的结果？还是手动继续做比如top 2*k的搜索来达到topk条？

**Reply by xiaobingxia-at** [reply]
> 每个segment各搜出topk条，n个segment就有n * topk条，这些都是按distance排好序的，很容易就能得到最终topk。很简单，每个高中高考前十名，排好序的，很容易得到全市的前十名。

这个是跨partition搜索是吧，不是上面那个关于index的问题的解释。

**Reply by liliu-z** [reply]
@xiaobingxia-at 
在Milvus上，我们目前支持了prefiltering，就像上边 @yhmo 描述那样。在索引里边，每个索引有不同的处理方式。比如HNSW，对无效点可以选择跳过去（不进candidate list，容易孤岛搜不够），也可以选择走过去（进candidate list，有缓解，容易速度慢，对太高过滤也会孤岛）。但是我们做了一些优化，比如超高过滤量转暴搜，比如是否进candidate list，进多少可调控，比如提升图质量降低局部链接等。
在ZillizCloud Cardinal上我们针对这个方向做了更多的调优，包括但不限于建立多层小图提高连通性，改变数据结构延缓孤岛，新的图算法，图、ivf等融合算法等方案。

同时我们计划在2.5.1支持Iterative-filtering，他会利用索引的ANN Iterator去拉取数据做过滤判断。这个是针对过滤量不高但是表达式比较复杂（bitset生成比较费劲）的场景。

**Reply by xiaobingxia-at** [reply]
@liliu-z 非常感谢您的详细解释。

> 比如超高过滤量转暴搜，比如是否进candidate list，进多少可调控，比如提升图质量降低局部链接等。

请问这些细节可以通过参数调节吗？
另外，对于用户来说，比如我找top10，但是过滤量太大（比如，整个数据集10m row，符合过滤条件的数据集只有2000行），这种我该如何找到top10的搜索结果呢？是直接搜top10000，然后在应用层过滤吗？谢谢、

**Reply by yhmo** [reply]
“整个数据集10m row，符合过滤条件的数据集只有2000行，找到top10的搜索结果” ---------------------------------- 这个就是普通的过滤后再计算topk就行，没有什么特别的。
假设一千万个小学生里面只有2000个是姓白的，那就只把这2000个姓白的小学生挑出来，找到最相似的10个即可。反映到具体调用上，就是 collection.search(expr="姓==白", limit=10)，根本不需要应用层做什么。

**Reply by xiaobingxia-at** [reply]
@yhmo 可能我没理解对。不是说，如果通过index，过滤量太大（过滤了9998000行），会有孤岛搜不够的情况吗？
我的理解是：
我从2000个姓白的找到最相似的10个，但是通过HNSWindex返回了10个最相似的但是姓张，然后因为都不姓白，就一条都没返回。
这是孤岛搜不够的情况吗？

**Reply by yhmo** [reply]
> @yhmo 可能我没理解对。不是说，如果通过index，过滤量太大（过滤了9998000行），会有孤岛搜不够的情况吗？ 我的理解是： 我从2000个姓白的找到最相似的10个，但是通过HNSWindex返回了10个最相似的但是姓张，然后因为都不姓白，就一条都没返回。 这是孤岛搜不够的情况吗？

开源的IVF索引和HNSW索引确实有这样的问题，不保证一定能搜出10个结果。对于IVF，你可以增加nprobe的值扩大搜索范围，当nprobe==nlist的时候，相当于暴搜，每条数据都被计算，就一定能搜到2000个姓白的；对于HNSW，你可以增加ef的值。扩大搜索范围就能搜出更多的结果，具体扩大多少，只能实测。

**Reply by liliu-z** [reply]
> @liliu-z 非常感谢您的详细解释。
> 
> > 比如超高过滤量转暴搜，比如是否进candidate list，进多少可调控，比如提升图质量降低局部链接等。
> 
> 请问这些细节可以通过参数调节吗？ 另外，对于用户来说，比如我找top10，但是过滤量太大（比如，整个数据集10m row，符合过滤条件的数据集只有2000行），这种我该如何找到top10的搜索结果呢？是直接搜top10000，然后在应用层过滤吗？谢谢、

目前这些参数应该没有暴露，使用经验值。在咱们这里例子里边，应该会自动转为暴搜提高recall和效率

**Reply by xiaobingxia-at** [reply]
@yhmo @liliu-z 再次感谢各位的详细解答。
最后确认一件事，如果过滤量比较大。假如没有转化为暴搜，还是通过HNSW索引搜索。最差结果只会出现搜不够的情况，不会出现召回率降低的情况（相比没有任何过滤的情况），我的理解对吗？谢谢。

**Reply by yhmo** [reply]
转为暴搜，意思是不走索引，因为大部分数据被过滤掉，计算量很小，就没必要再走索引，过滤出1000条就计算1000条，过滤出2000条就计算2000条。

**Reply by liliu-z** [reply]
> @yhmo @liliu-z 再次感谢各位的详细解答。 最后确认一件事，如果过滤量比较大。假如没有转化为暴搜，还是通过HNSW索引搜索。最差结果只会出现搜不够的情况，不会出现召回率降低的情况（相比没有任何过滤的情况），我的理解对吗？谢谢。

这个recall的ground truth也是过滤之后暴搜的结果，recall会下降。比如搜topk，这k个点都在图的一个角落，entrypoint进去后到这个角落的路径上的点都被过滤了。哪怕选择踩过去，由于ef(size of the candidate list)限制，会导致没到这个角落就返回了，recall为0。 扩大ef有助于缓解

**Reply by xiaobingxia-at** [reply]
@liliu-z 目前索引转暴力搜索的“比例”是多少？比如filter的数据量达到多少百分比会变成暴力搜索？

> 哪怕选择踩过去，由于ef(size of the candidate list)限制，会导致没到这个角落就返回了，recall为0。 扩大ef有助于缓解

嗯嗯，我的表达可能不太准确。应该说，要么就是返回的结果不够k个。但一旦返回了，这个返回的结果就应该和正常HNSW无过滤搜索的结果是一致的。对吗？

**Reply by xiaofan-luan** [reply]
这个是不一定，我们基本能保证k个结果返回，但是结果不一致时很可能的。比如某一个点，所有的邻居都被包围了，就可能会漏结果

**Reply by xiaobingxia-at** [reply]
@xiaofan-luan 不好意思，我再补充一下。
比如，无过滤用HNSW，返回的结果是1,2,3,4,5
有过滤用HNSW，返回的结果可能是1,2,3 （漏掉了4,5），但是无论怎样不会出现返回6,7,8的情况。这个理解对吗？
然后，提高ef值，可能会从返回1,2,3变成1,2,3,4,5，但是会更慢。

**Reply by liliu-z** [reply]
> @xiaofan-luan 不好意思，我再补充一下。 比如，无过滤用HNSW，返回的结果是1,2,3,4,5 有过滤用HNSW，返回的结果可能是1,2,3 （漏掉了4,5），但是无论怎样不会出现返回6,7,8的情况。这个理解对吗？ 然后，提高ef值，可能会从返回1,2,3变成1,2,3,4,5，但是会更慢。

是的

**Reply by HowardQin** [reply]
> > @xiaofan-luan 不好意思，我再补充一下。 比如，无过滤用HNSW，返回的结果是1,2,3,4,5 有过滤用HNSW，返回的结果可能是1,2,3 （漏掉了4,5），但是无论怎样不会出现返回6,7,8的情况。这个理解对吗？ 然后，提高ef值，可能会从返回1,2,3变成1,2,3,4,5，但是会更慢。
> 
> 是的

@liliu-z @xiaobingxia-at 关键是加了过滤后，符合条件的记录也可能因为被隔离而漏掉，为什么123不可能被隔离而漏掉呢？我觉得返回678的情况是有可能的，除非你能证明如果能返回678就一定能返回123。

**Reply by yhmo** [reply]
> > > @xiaofan-luan 不好意思，我再补充一下。 比如，无过滤用HNSW，返回的结果是1,2,3,4,5 有过滤用HNSW，返回的结果可能是1,2,3 （漏掉了4,5），但是无论怎样不会出现返回6,7,8的情况。这个理解对吗？ 然后，提高ef值，可能会从返回1,2,3变成1,2,3,4,5，但是会更慢。
> > 
> > 
> > 是的
> 
> @liliu-z @xiaobingxia-at 关键是加了过滤后，符合条件的记录也可能因为被隔离而漏掉，为什么123不可能被隔离而漏掉呢？我觉得返回678的情况是有可能的，除非你能证明如果能返回678就一定能返回123。

HNSW是图索引，打个比方就类似于在中国地图上找符合某个条件的小镇，并且只给你一箱汽油，没油了就结束。
比如，我想找10个人口大于10万的镇子，我寻思东南部的镇子人口应该比较密集，所以就朝着东南面开车找，碰到123镇子，它们的人口数都大于10万，ok是我要找的，然后我又继续开车遇到678，它们的人口小于10万，那么它们肯定被过滤掉，虽然我路过了它们但我可以无视。最后我车没油了，45没找到，因为45在西北面啊，我遇不到它们。那么最后返回的就是123
